src/charm.py

```diff
@@ -431,6 +441,103 @@ def _on_new_vault_kv_client_attached(self, event: NewVaultKvClientAttachedEvent)
         stale_nonces = set(credential_nonces) - set(nonces)
         self.vault_kv.remove_unit_credentials(relation, stale_nonces)
 
+    def _on_create_backup_action(self, event: ActionEvent) -> None:
+        """Handles create-backup action.
+
+        Creates backup and stores it in S3 bucket.
+
+        Args:
+            event: ActionEvent
+        """
+        if not self._is_relation_created(S3_RELATION_NAME):
+            event.fail(message="S3 relation not created. Failed to perform backup operation.")
+            return
+
+        if not self.unit.is_leader():
+            event.fail(message="Only leader unit can perform backup operations.")
+            return
+
+        missing_parameters = self._get_missing_s3_parameters()
+        if missing_parameters:
+            event.fail(message=f"S3 parameters missing. {missing_parameters}")
+            return
+
+        s3_parameters = self._retrieve_s3_parameters()
+
+        if not (session := self._create_s3_session(s3_parameters)):
+            event.fail(message="Failed to create S3 session.")
+            return
+
+        try:
+            self._create_s3_bucket(
+                session=session,
+                region=s3_parameters["region"],
+                bucket_name=s3_parameters["bucket"],
+                endpoint=s3_parameters["endpoint"],
+            )
+            logger.info("Created S3 bucket %s", s3_parameters["bucket"])
+        except KeyError as e:
+            logger.warning("Missing required S3 parameter: %s", e)
```

Comment:
I think we should log at the `error` level here, we are failing the action right after.

Reply:
✅